### PR TITLE
Add support for AES-CMAC according to RFC 4493

### PR DESCRIPTION
**Pull request by doak**
This extends the existing generic `MessageAuthenticationCode` interface
with a second implementation besides HMAC. Given that the existing
naming suggests it this way, both implementations reside in `mac.cpp`
for now, but could be separated in future work.

The public interfaces are only extended, i.e. this change retains
API/ABI compatibility.

### CODE REVIEW

File: src/mococrw/mac.h
Comment (neverpanic) [inline src/mococrw/mac.h]:
```suggestion
     * @param cipherType The type of encryption function which shall be used
```

File: src/mococrw/mac.h
Comment (neverpanic) [inline src/mococrw/mac.h]:
```suggestion
     * Adds the chunk of data to the MAC.
```

I believe `H(i-key || message1 [ ||  message2 ])` is the formula for HMAC, is it not?

Also, hashing isn't involved in AES-CMAC, so let's not use the word "hash".

File: src/mococrw/mac.h
Comment (neverpanic) [inline src/mococrw/mac.h]: This also applies here, this is how HMAC is computed.

File: src/mococrw/openssl_wrap.h
Comment (neverpanic) [inline src/mococrw/openssl_wrap.h]: I know that should have happened already, but can you please `#include <limits>` for `std::numeric_limits`?

File: src/openssl_wrap.cpp
Comment (neverpanic) [inline src/openssl_wrap.cpp:1432]: Can we statically guarantee that this will never happen? Otherwise it may be better to call `CMAC_Final` twice, once with a `nullptr` as buffer to obtain the required length, and then with a buffer of the required size.

File: tests/unit/test_cmac.cpp
Comment (neverpanic) [inline tests/unit/test_cmac.cpp:1]: Should we also test that invoking `update()` multiple times works as expected?

**Review by neverpanic — CHANGES_REQUESTED** [review]
A couple of comments.

File: src/openssl_wrap.cpp
Comment (doak) [reply src/openssl_wrap.cpp:1432]: There is even an API available (`EVP_CIPHER_CTX_block_size()`) to get the block size of a specific cipher.
But yes, MAC size is specified as the block size and `EVP_MAX_BLOCK_LENGTH` represents the biggest block size of any cipher supported by OpenSSL. Looking into the OpenSSL source of `CMAC_Final()` also proves that this should work.

Do you prefer to use `EVP_CIPHER_CTX_block_size()`?

File: tests/unit/test_cmac.cpp
Comment (doak) [reply tests/unit/test_cmac.cpp:1]: I would not expect any issues here, but sure, why not.
I will also note that on the list of potentiall improvements for the HMAC implementation.

File: src/mococrw/mac.h
Comment (doak) [reply src/mococrw/mac.h]: Yes, sorry. I did miss the comments of this header :(

File: src/mococrw/openssl_wrap.h
Comment (doak) [reply src/mococrw/openssl_wrap.h]: Yep.
On the other hand nothing uses this at all. Do you mind just removing `NONE` and, if your linter on CI is good enough, perhaps also remove the `default` case in the mapper `_getCipherPtrFromCmacCipherType()`?
(I have just used that construct because it was done that way for HMAC.)

File: src/mococrw/mac.h
Comment (doak) [reply src/mococrw/mac.h]: Nice! I have just saw that the CI failed because of the undocumented parameter :)

**Comment by doak** [comment]
Regarding doxygen comments: I just saw that a there is a mixture of Doxygen comments with and without trailing full stop (copied over from HMAC). I will remove the trailing full stops as well.

File: src/mococrw/openssl_wrap.h
Comment (neverpanic) [reply src/mococrw/openssl_wrap.h]: The problem with doing that is that it would be inconsistent to not have it for CmacCipherTypes but keep it for HMAC, but removing it from HMAC is an ABI and API change. Let's just keep it.

File: src/openssl_wrap.cpp
Comment (neverpanic) [reply src/openssl_wrap.cpp:1432]: No, this is fine as-is then.

File: src/mococrw/openssl_wrap.h
Comment (doak) [reply src/mococrw/openssl_wrap.h]: Done.

**Review by neverpanic — APPROVED** [review]
lg2m now, but I'd like to wait for a second review.

File: src/mococrw/mac.h
Comment (doak) [reply src/mococrw/mac.h]: Done.

File: tests/unit/test_cmac.cpp
Comment (doak) [reply tests/unit/test_cmac.cpp:1]: Done.

**Comment by doak** [comment]
I took the opportunity to harmonise Doxygen comments throughout `mococrw/mac.h` (including existing ones => dedicated commit) as well as to split up the test of the control flow into several test cases (only for CMAC).

**Review by neverpanic — APPROVED** [review]

**Comment by neverpanic** [comment]
Unfortunately some of our internal systems use older copies of gtest, which seems to cause this issue:
```
/usr/bin/c++   -I/src/src -isystem /usr/src/gmock/include -isystem /usr/src/gmock -isystem /usr/src/googletest/include -isystem /usr/src/googletest -fPIE   -Wall -Wextra -pedantic -fstack-protector-strong -Wfatal-errors -pthread -std=gnu++14 -MD -MT tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o -MF tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o.d -o tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o -c /src/tests/unit/test_cmac.cpp
/src/tests/unit/test_cmac.cpp:155:25: error: expected constructor, destructor, or type conversion before '(' token
 INSTANTIATE_TEST_SUITE_P(
                         ^
compilation terminated due to -Wfatal-errors.
```

Additionally, it seems this doesn't compile with clang:
```
/usr/lib/ccache/clang++   -I/src/src -isystem /usr/src/gmock/include -isystem /usr/src/gmock -isystem /usr/src/googletest/include -isystem /usr/src/googletest -fsanitize=address -fno-omit-frame-pointer -fsanitize-address-use-after-scope -g -O1 -fPIE   -Wall -Wextra -pedantic -fstack-protector-strong -Wfatal-errors -pthread -std=gnu++14 -MD -MT tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o -MF tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o.d -o tests/unit/CMakeFiles/cmactests.dir/test_cmac.cpp.o -c /src/tests/unit/test_cmac.cpp
/src/tests/unit/test_cmac.cpp:156:9: fatal error: unknown type name 'cmac'
        cmac,
        ^
```

This may be the same issue with gtest, since the location is similar, and the difference may just be in the error messages printed by the compilers.

**Comment by doak** [comment]
@neverpanic, I am sorry. I did not take into account that you could use an older version of _gtest_. It is marked as deprecated for my local version.
Fixed that. Yes: the second issue is very likley caused by the first.

**Review by neverpanic — APPROVED** [review]
Thanks that solves the issue. We have a centrally maintained copy of gtest that's a bit dated.

File: src/mac.cpp
Comment (garw) [inline src/mac.cpp]: The AES Crypto implementation does a test on the key size and provides a MoCOCrW exception with an easy to understand error message  if the sizes don't add up. We should probably do the same here.

File: src/mac.cpp
Comment (garw) [inline src/mac.cpp]: Any particular reasons this is a shared_ptr instead of a unique_ptr?

File: src/mac.cpp
Comment (garw) [inline src/mac.cpp]: isn't this the default as well?

File: src/mococrw/mac.h
Comment (garw) [inline src/mococrw/mac.h:159]: In the current implementation, this throws if cipherType is NONE. I wonder if we should really add this NONE type. This way, we could syntactically ensure that this call will never fail due to an invalid cipher.

The documentation should also reflect that it throws a MoCOCrWException if the key size is wrong.

File: src/mococrw/openssl_wrap.h
Comment (garw) [reply src/mococrw/openssl_wrap.h]: Honestly, I kind of agree that not having NONE is better design. I fear we have this implict optional build-in at other places for digests but I don't think that its a good idea to continue this. We should rather reconsider the NONE on DigestTypes in the future…

File: tests/unit/test_cmac.cpp
Comment (garw) [inline tests/unit/test_cmac.cpp]: Strictly speaking this opens the door to undefined behavior. Maybe we want to be a good example and do incorrectCmac[0] ^= 0x1;

**Review by garw — CHANGES_REQUESTED** [review]
Looks mostly good. A few knitpicks regarding some design issues.

File: src/mac.cpp
Comment (doak) [reply src/mac.cpp:187]: No hard rationale. It is done the same within the HMAC implementation and, since raw pointers are used for the next level of abstraction (within `CMAC::Impl`), I thought a `shared_ptr` reflect this fact the best.

Btw., my first intention was to extract the raw pointer as late as possible just for the calls of the `SSL_*` methods (i.e. within call of `*::callChecked()` in `openssl_wrap.cpp`), but I decided against it because it is used differently all over the place in the existing code. (I am not sure yet how much we are allowed to sacrifice consistency in favour of cleaner code.)

File: tests/unit/test_cmac.cpp
Comment (doak) [reply tests/unit/test_cmac.cpp]: I appreciate that.
Just to sort it out correctly for me: This particular case is well defined (unsigned integral type) but to avoid potentially problems, such a pattern should be avoided if possible, right?

File: src/mac.cpp
Comment (doak) [reply src/mac.cpp]: No hard rationale. It is done the same within the HMAC implementation and, since raw pointers are used for the next level of abstraction (within `CMAC::Impl`), I thought a `shared_ptr` reflect this fact the best.

Btw., my first intention was to extract the raw pointer as late as possible just for the calls of the `SSL_*` methods (i.e. within call of `*::callChecked()` in `openssl_wrap.cpp`), but I decided against it because it is used differently all over the place in the existing code. (I am not sure yet how much we are allowed to sacrifice consistency in favour of cleaner code.)

File: src/mococrw/mac.h
Comment (doak) [reply src/mococrw/mac.h:159]: I just had another look into the code including OpenSSL itself. In fact `openssl::DigestTypes::NONE` is used e.g. for `EdDSASignaturePublicKeyCtx::verifyMessage()` to allow OpenSLL to select a default digest by its own.
Regardless that `NONE` is quite a bad name for his behaviour, there seems to be a valid use case for this.

Such a behaviour does not exist for AES-CMAC (in current OpenSSL implementation at least).
I guess you still vote for removing `openssl::CmacCipherTypes::None`, right?

File: src/mac.cpp
Comment (garw) [reply src/mac.cpp]: Then, I'd suggest to move to unique_ptr here. This is no public interface. We even change it to unique_ptr in the hmac without impacting anyone due to the impl pattern. Can you just change it in both places?

File: src/mococrw/mac.h
Comment (garw) [reply src/mococrw/mac.h:159]: Yes. Let's try to (slowly but constantly) move towards a better code-base.

File: tests/unit/test_cmac.cpp
Comment (garw) [reply tests/unit/test_cmac.cpp]: A true, I didn't check that its unsigned. So you're right. Maybe we want to change this anyways?

Comment (doak) [reply tests/unit/test_cmac.cpp]: Sure.

File: src/mococrw/openssl_wrap.h
Comment (doak) [reply src/mococrw/openssl_wrap.h]: Removed `None` from `CmacCipherTypes`.

File: src/mac.cpp
Comment (doak) [reply src/mac.cpp]: Done.

File: src/mac.cpp
Comment (doak) [reply src/mac.cpp]: Done.

File: src/mac.cpp
Comment (doak) [reply src/mac.cpp]: Done.

File: src/mococrw/mac.h
Comment (doak) [reply src/mococrw/mac.h:159]: Exception with message is thrown. Documentation and test case updated.

File: tests/unit/test_cmac.cpp
Comment (doak) [reply tests/unit/test_cmac.cpp]: Done.

**Review by garw — APPROVED** [review]

**Review by neverpanic — APPROVED** [review]

**Pull request merged**
